### PR TITLE
Assemble_coupling_exact

### PR DESCRIPTION
**Pull request by fdrmrc**
This computes <v_i,q_j>_{B} where B is a domain embedded in \Omega.  It exploits a vector of tuples to distribute the local contributions, which is supposed to be defined and computed using `compute_intersection(space_cache, immersed_cache,degree)` in user codes. 
All in all:

        const auto cells_with_quads =
          NonMatching::compute_intersection(*space_grid_tools_cache,
                                            *embedded_grid_tools_cache,
                                            degree);

        NonMatching::assemble_coupling_exact(space_dh,
                                             embedded_dh,
                                             cells_with_quads,
                                             coupling_matrix,
                                             constraints,
                                             ComponentMask(),
                                             ComponentMask(),
                                             *space_mapping,
                                             embedded_mapping(),
                                             embedded_constraints);

Checked with `reduced_lagrange.cc` with `dim=spacedim=2`. If you don't mind, I'll add a test with it in the Testsuite. I moved the embedded grid a little so to have non-trivial intersections:
 
distributed_codim_0

TODO:

- [x] ComponentMask(s) 
- [x] Tests(?).

### CODE REVIEW

**Comment by luca-heltai** [comment]
When you add tests for things that will go in the deal.II library (like this one), use the standard deal.II tests way of doing things (modify `dealiii_tests/template.cc`). Make a test that produces a "golden output", run it, and save the `output` file to `nome_test_XX.output`.

File: source/create_coupling_sparsity_pattern_with_exact_intersections.cc [hunk 1]
```diff
@@ -108,8 +115,16 @@ namespace dealii::NonMatching
           }
       }
 
+    const bool dof_mask_is_active =
+      dof_mask.n_rows() == n_space_dofs && dof_mask.n_cols() == n_immersed_dofs;
+    Assert(
```
Comment (fdrmrc) [inline source/create_coupling_sparsity_pattern_with_exact_intersections.cc:120]: I had to add this Assert for the case when `dof_mask` is active and I have non-trivial masks, as it seems to me that it's currently not implemented ( seehttps://github.com/dealii/dealii/blob/c487e7fe81417c23c3d8d4dff437665f9a827916/include/deal.II/lac/affine_constraints.templates.h#L4406)

**Comment by luca-heltai** [comment]
take a look at the way we implemented this in the `assemble_coupling` that is in deal.II now. There is a TODO note there, and a "slow" way to assemble those, looping for all i and j dofs, and distributing them one by one.

**Comment by fdrmrc** [comment]
I have that "slow way" in my current assembly, but the problem here is in the sparsity pattern. Are you suggesting to use that way to fill the sparsity, too? If so we don't need to use`dof_mask`.

**Comment by luca-heltai** [comment]
Yes, we should use that also for the sparsity, only when masks are provided that are not default ones, as in your if else block.

**Comment by luca-heltai** [comment]
Can you remove the "draft" option when this is ready?

**Pull request merged**